src/shmem_lock.c

```diff
@@ -0,0 +1,173 @@
+/* -*- C -*-
+ *
+ * Copyright 2011 Sandia Corporation. Under the terms of Contract
+ * DE-AC04-94AL85000 with Sandia Corporation, the U.S.  Government
+ * retains certain rights in this software.
+ *
+ * Copyright (c) 2018 Intel Corporation. All rights reserved.
+ * This software is available to you under the BSD license.
+ *
+ * This file is part of the Sandia OpenSHMEM software package. For license
+ * information, see the LICENSE file in the top level directory of the
+ * distribution.
+ *
+ */
+
+#include <stdlib.h>
+#include <pthread.h>
+#include <stdint.h>
+#include "shmem_internal.h"
+#include "shmem_lock.h"
+#include "uthash.h"
+
+typedef struct {
+    UT_hash_handle  hh;
+    long           *key;
+    uint64_t        next_ticket;
+    uint64_t        cur_ticket;
+    pthread_mutex_t mutex;
+    pthread_cond_t  cond[];
+} shmem_internal_lock_guard_t;
+
+static shmem_internal_lock_guard_t *guards = NULL;
+
+/* Simple queueing lock using Lamport's bakery algorithm.  Uses LOCK_QUEUE_SIZE
+ * condition variables per lock to reduce the number of threads that are woken
+ * up when the lock is released. */
+static inline void shmem_internal_qlock_lock(shmem_internal_lock_guard_t *g) {
+    int ret;
+    char errmsg[256];
+    uint64_t my_ticket;
+
+    ret = pthread_mutex_lock(&g->mutex);
+    if (ret) RAISE_ERROR_MSG("pthread_mutex_lock failed: %s\n",
+                             shmem_util_strerror(ret, errmsg, 256));
+
+    my_ticket = g->next_ticket++;
```

Comment:
Does this (and others below) need to be an atomic read and inc?

Reply:
Nope, it's guarded by `g->mutex`.

Reply:
Doesn't the `pthread_cond_wait` release the mutex?

Reply:
`pthread_cond_wait` also reacquires the mutex before it returns.